Classes/Promise.swift

```diff
@@ -0,0 +1,269 @@
+//
+//  BAPromise.swift
+//  OS X
+//
+//  Created by Ben Allison on 6/15/18.
+//  Copyright © 2018 Ben Allison. All rights reserved.
+//
+
+import Foundation
+
+public enum PromiseResult<ValueType> {
+    case success(ValueType)
+    case promise(Promise<ValueType>)
+    case failure(Error)
+    
+    var resolved: Bool {
+        switch(self) {
+        case .success, .failure:
+            return true
+        default:
+            return false
+        }
+    }
+    
+    init(error: Error) {
+        self = .failure(error)
+    }
+    
+    init(promise: Promise<ValueType>) {
+        self = .promise(promise)
+    }
+    
+    init(value: ValueType) {
+        self = .success(value)
+    }
+    
+}
+
+extension PromiseResult where ValueType == Void {
+    static var success: PromiseResult {
+        return .success(())
+    }
+    
+    init() {
+        self = .success(())
+    }
+}
+
+public class PromiseCancelToken {
+    var cancelled: Bool = false
+    static let queue = DispatchQueue(label: "com.github.benski.promise")
+    var onCancel : (() -> Void)?
+    
+    func cancelled(_ onCancel: @escaping () -> Void, on queue: DispatchQueue) {
+        let wrappedBlock = {
+            queue.async {
+                onCancel()
+            }
+        }
+        
+        PromiseCancelToken.queue.async {
+            if self.cancelled {
+                wrappedBlock()
+            } else {
+                self.onCancel = wrappedBlock
+            }
+        }
+    }
+    
+    func cancel() {
+        cancelled = true
+        OSMemoryBarrier()
```

Comment:
Why do we need it here? 
as far as I see OSMemoryBarrier is deprecated. 
`@available(iOS, deprecated: 10.0, message: "Use atomic_thread_fence() from <stdatomic.h> instead")`